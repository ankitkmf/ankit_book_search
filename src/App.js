import React, { Component } from "react";
import Axios from "axios";
import AllResults from "./components/AllResults";
import SingleResult from "./components/SingleResult";
import FacebookLoading from 'react-facebook-loading';
import 'react-facebook-loading/dist/react-facebook-loading.css';

const apiKey ='sYkovQcCNOGAnLuiCTEZrQ';// Create .env file and store in environment variable

class App extends Component {
  state = {
    searchText: "",
    error: "",
    fetchingData: false,
    searchResults: [],
    singleResult: [],  
    showAllData:true
  };

  onTextChange = e => {
    this.setState({
      searchText: e.target.value
    });
  };

  onSearchBookClick = () => {
    this.setState({fetchingData: true});
    const { searchText } = this.state;
    const requestUri = `https://cors-anywhere.herokuapp.com/https://www.goodreads.com/search/index.xml?key=${apiKey}&q=${searchText}`;

    Axios.get(requestUri)
      .then(res => {
        this.parseXMLResponse(res.data);
      })
      .catch(error => {
        this.setState({
          error: error.toString(),
          fetchingData: false
        });
      });

      this.setState({showAllData:true});
  };

 
  parseXMLResponse = response => {
    const parser = new DOMParser();
    const XMLResponse = parser.parseFromString(response, "application/xml");
    const parseError = XMLResponse.getElementsByTagName("parsererror");

    if (parseError.length) {
      this.setState({
        error: "There was an error fetching results.",
        fetchingData: false
      });
    } else {
     
      if (this.state.showAllData)
      {
      
        const XMLresults = new Array(...XMLResponse.getElementsByTagName("work"));
        const searchResults = XMLresults.map(result => this.XMLToJson(result));
        this.setState({ searchResults, fetchingData: false });
      }
      else
      {
       
        const XMLresults = new Array(...XMLResponse.getElementsByTagName("book"));
        const searchResults = XMLresults.map(result => this.XMLToJson(result));
        this.setState({ singleResult: searchResults, fetchingData: false });
      }
    }
  };


  XMLToJson = XML => {
    const allNodes = new Array(...XML.children);
    const jsonResult = {};
    allNodes.forEach(node => {
      if (node.children.length) {
        jsonResult[node.nodeName] = this.XMLToJson(node);
      } else {
        jsonResult[node.nodeName] = node.innerHTML;
      }
    });
    return jsonResult;
  };

  getSingleBook=id=>{
    this.setState({fetchingData: true});
    const requestUri =`https://cors-anywhere.herokuapp.com/https://www.goodreads.com/book/show/${id}?key=${apiKey}`;      
    Axios.get(requestUri)
      .then(res => {
        this.setState({showAllData: false });      
        this.parseXMLResponse(res.data);
      })
      .catch(error => {    
        this.setState({
          error: error.toString(),
          fetchingData: false
        });      
      });    
  }

  resetView=()=>{
    this.setState({showAllData: true });
  }

  render() {
    return (
      <div className="container">
        <div className="header clearfix mt-5">
          <h3 className="text-muted">Ankit Book Search</h3>
        </div>
        <div className="jumbotron">
          <div className="form-group row">
            <input
              className="mr-1 col-sm-9 form-control"
              type="text"
              placeholder="Search Books By title, author"
              name="searchText"
              onChange={this.onTextChange}
              value={this.state.searchText}
            />
            <button
              className="col-sm-2 btn btn-primary"
              onClick={this.onSearchBookClick}
            >
              Search
            </button>
          </div>

          {this.state.fetchingData ? (           
            <FacebookLoading
              delay={0}
              duration={800}
              zoom={4}
              style={{ margin: '0 auto' }}
            />
          ) : (
              (this.state.error && (
                <p className="text-danger">{this.state.error}</p>
              )) || 
              
              (this.state.showAllData ? 
                <AllResults books={this.state.searchResults} getSingleBook={this.getSingleBook.bind(this)}/>
                :  <SingleResult books={this.state.singleResult} resetView={this.resetView.bind(this)}/>
              )
            )}
        </div>
      </div>
    );
  }
}

export default App;
