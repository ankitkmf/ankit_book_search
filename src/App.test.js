import React from 'react';
import App from './App';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow, render, mount } from 'enzyme';
import renderer from 'react-test-renderer';


Enzyme.configure({ adapter: new Adapter() });

// test("fake test",()=>{
//   expect(true).toBeTruthy() ;
// });

it('renders without crashing', () => {
  shallow(<App />);
});

describe("<App/>", () => {
  it("render an serch area", () => {
    const component = shallow(<App />);
    expect(component.find('input').length).toEqual(1);
  })

  it("render an button", () => {
    const component = shallow(<App />);
    expect(component.find('button').length).toEqual(1);
  })

  it("render an h3 test", () => {
    const component = shallow(<App />);
    const text = component.find('h3.text-muted').text();
    expect(text).toEqual("Ankit Book Search");

  }) 

});

describe("<App/> snapshot", () => {
  it('App snapshot renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});