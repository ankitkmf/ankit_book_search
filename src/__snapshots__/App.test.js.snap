// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> snapshot App snapshot renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="header clearfix mt-5"
  >
    <h3
      className="text-muted"
    >
      Ankit Book Search
    </h3>
  </div>
  <div
    className="jumbotron"
  >
    <div
      className="form-group row"
    >
      <input
        className="mr-1 col-sm-9 form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Search Books By title, author"
        type="text"
        value=""
      />
      <button
        className="col-sm-2 btn btn-primary search"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
    <div
      className="row allResult"
    />
  </div>
</div>
`;
